pkg/agent/multicluster/mc_route_controller.go

```diff
@@ -436,10 +634,50 @@ func generatePeerConfigs(subnets []string, gatewayIP net.IP) (map[*net.IPNet]net
 	return peerConfigs, nil
 }
 
-// getPeerGatewayIP will always return the first Gateway IP.
-func getPeerGatewayIP(spec mcv1alpha1.ClusterInfo) net.IP {
+// If WireGuard is disabled, getPeerGatewayTunnelIP will return Gateway's GatewayIP.
+// If WireGuard is enabled, the WireGuard interfaces use the first IP address of ServiceCIDR
+// as its IP address. So getPeerGatewayTunnelIP will return the first IP of the ServiceCIDR
+// as the remote Gateway tunnel IP.
+func getPeerGatewayTunnelIP(spec mcv1alpha1.ClusterInfo, enableWireGuard bool) net.IP {
+	if enableWireGuard {
+		if spec.ServiceCIDR == "" {
+			klog.InfoS("The ServiceCIDR has not been updated, skip.", "clusterID", spec.ClusterID)
+			return nil
+		}
+		_, serviceCIDR, err := net.ParseCIDR(spec.ServiceCIDR)
+		if err != nil {
+			klog.Warningf("Failed to get the peer Gateway IP for WireGuard, error: %v", err)
+			return nil
+		}
+		return serviceCIDR.IP
+	}
 	if len(spec.GatewayInfos) == 0 {
 		return nil
 	}
 	return net.ParseIP(spec.GatewayInfos[0].GatewayIP)
 }
+
+func getLocalGatewayIP(gateway *mcv1alpha1.Gateway, enableWireGuard bool) net.IP {
+	if enableWireGuard {
+		if gateway.ServiceCIDR == "" {
+			klog.InfoS("The ServiceCIDR has not been updated, skip.", "Gateway", klog.KObj(gateway))
+			return nil
+		}
+		localGatewayIP, _, err := net.ParseCIDR(gateway.ServiceCIDR)
+		if err != nil {
+			klog.Warningf("Failed to get the local Gateway IP for WireGuard, error: %v", err)
+			return nil
+		}
+		return localGatewayIP
+	}
+	return net.ParseIP(gateway.GatewayIP)
+}
+
+// isWireGuardInfoChanged checks the information in ClusterInfoImport needed by WireGuard change or not.
+func isWireGuardInfoChanged(cache, cur *mcv1alpha1.ClusterInfoImport) bool {
+	if cache.Spec.ServiceCIDR == cur.Spec.ServiceCIDR && cache.Spec.WireGuard != nil && cur.Spec.WireGuard != nil &&
+		cache.Spec.WireGuard.PublicKey == cur.Spec.WireGuard.PublicKey {
+		return false
+	}
+	return true
```

Comment:
It should return false when both old WireGuard and new WireGuard are nil.
```suggestion
        if cache.Spec.ServiceCIDR != cur.Spec.ServiceCIDR {
                return true
        }
	if cache.Spec.WireGuard == nil && cur.Spec.WireGuard == nil {
	        return false
	}
        if cache.Spec.WireGuard == nil || cur.Spec.WireGuard == nil {
	        return true
	}
	return cache.Spec.WireGuard.PublicKey != cur.Spec.WireGuard.PublicKey
```